extra_data/write_cxi.py

```diff
@@ -38,31 +44,50 @@ def __init__(self, detdata):
             frame_counts.index.values, frame_counts.values.astype(np.intp)
         )
 
-        # cumulative sum gives the end of each train, subtract to get start
+        # Cumulative sum gives the end of each train, subtract to get start
         self.train_id_to_ix = frame_counts.cumsum() - frame_counts
 
     @property
     def nmodules(self):
         return self.detdata.n_modules
 
+    @property
+    def ncells(self):
+        return 1
+
     @property
     def data(self):
         return self.detdata.data
 
     def collect_pulse_ids(self):
+        """
+        Gather pulse/cell ID labels for all modules and check consistency.
+
+        Raises
+        ------
+
+        Exception:
+          Some data has no pulse ID values for any module.
+        Exception:
+          Inconsistent pulse IDs between detector modules.
+
+        Returns
+        -------
+
+        pulse_ids_min: np.array
+          Array of pulse IDs per frame common for all detector modules.
+        """
         # Gather pulse IDs
         NO_PULSE_ID = 9999
         pulse_ids = np.full((self.nframes, self.nmodules), NO_PULSE_ID,
                             dtype=np.uint64)
 
         for source, modno in self.detdata.source_to_modno.items():
-            for chunk in self.data._find_data_chunks(source, 'image.pulseId'):
-                # In some cases, there's an extra dimension of length 1
-                if chunk.dataset.ndim > 1:
-                    chunk_data = chunk.dataset[:, 0]
-                else:
-                    chunk_data = chunk.dataset
-                self._map_chunk(chunk, chunk_data, pulse_ids, modno)
+            module_id = self.modulenos.index(modno)
+            for chunk in self.data._find_data_chunks(source,
+                                                     self.pulse_id_label):
+                chunk_data = chunk.dataset
+                self._map_chunk(chunk, chunk_data, pulse_ids, module_id)
```

Comment:
I don't think changing `modno` to `module_id` here is correct. If module 4 is missing from our data and we exclude it, then we should leave a gap at index 4, not shift module 5 to that location.

Looking through the code, I think we've always done the wrong thing for the `module_identifier` dataset in that case - we presumably haven't noticed because CrystFEL doesn't use that.

Reply:
Implemented in https://github.com/European-XFEL/EXtra-data/commit/cc73dadb190a2c3e831a1fb08259f2752871ee5d. Please let me know what should we do with `module_identifier`? Right now it looks fine only if all modules are present in the data.

Reply:
Thanks. I think `module_identifier` should be like `list(range(n_modules))` for the xtdf detectors, and 1 ... N+1 for JUNGFRAU, i.e. number all of the modules which there's 'space' for, even the ones that aren't included.

Reply:
Thank you for the suggestion, implemented in https://github.com/European-XFEL/EXtra-data/commit/2758b23ae00008c8ca120cdc6ee6ae01cb6a04dd.